src/addon/manager_ui.cpp

```diff
@@ -276,8 +278,29 @@ bool ad_hoc_addon_fetch_session(const std::vector<std::string>& addon_ids)
 			addons_list::const_iterator it = addons.find(addon_id);
 			if(it != addons.end()) {
 				const addon_info& addon = it->second;
-				// don't redownload in case it was already downloaded for being another add-on's dependency
-				if(!filesystem::file_exists(filesystem::get_addons_dir()+"/"+addon_id)) {
+				const std::string addon_dir = filesystem::get_addons_dir()+"/"+addon_id;
+				const std::string info_cfg = addon_dir+"/_info.cfg";
+
+				// no _info.cfg, so either there's a _server.pbl or there's no version information available at all, so this add-on can be skipped
+				if(filesystem::file_exists(addon_dir) && !filesystem::file_exists(info_cfg)) {
+					INFO_CFG << "No _info.cfg exists for '" << addon_id << "', skipping update.\n";
+					continue;
+				}
+
+				// if _info.cfg exists, compare the local vs remote add-on versions to determine whether a download is needed
+				// else, the add-on doesn't exist locally and needs to be downloaded
+				if(filesystem::file_exists(info_cfg)) {
+					game_config::config_cache& cache = game_config::config_cache::instance();
+					config info;
+					cache.get_config(info_cfg, info);
+					version_info installed_addon_version(info.child_or_empty("info")["version"]);
+
+					// if the installed version is outdated, download the most recent version from the add-ons server
+					if(installed_addon_version < addon.current_version) {
+						addons_client::install_result res = client.install_addon_with_checks(addons, addon);
+						return_value = return_value && (res.outcome == addons_client::install_outcome::success);
```

Comment:
The contents of this conditional block are the same as the contents of the next conditional block, it would be nice to merge them.

The logic above has already used a `continue` to avoid an unnecessary download, so I think the logic here should use `continue` too.

Reply:
Did the first part, though I'm not sure what you mean about the `continue`?

Reply:
I mean that the logic could do this:
```
if(...)
    continue;
if(...)
    continue;
addons_client::install_stuff();
```
or it could do this:
```
bool do_update=true;
if(...)
    do_update=false;
if(...)
    do_update=false;
if(do_update==false)
    addons_client::install_stuff();
```
but please choose one or the other without mixing them.

Reply:
The point of it is simply to short circuit things when there is no chance of doing either an install or an update. I could add a `do_nothing` variable and then change the if-checks that come after it to also check that variable, but that would then also mean the last part of the current logic - where it displays any add-ons which needed an install/update but weren't found on the add-ons server - would never trigger anymore.

Reply:
Maybe I'm missing something - I'm not seeing why you do this:
```
// if the installed version is outdated, download the most recent version from the add-ons server
if(installed_addon_version < addon.current_version) {
    do_update = true;
}
```
instead of doing
```
// if the installed version is already up-to-date, skip downloading a new copy from the add-ons server
if(installed_addon_version >= addon.current_version) {
    continue;
}
```

Reply:
Removing `do_update` and using `continue` would mean that the check immediately below it would be just:
```
if(!filesystem::file_exists(info_cfg)) {
	addons_client::install_result res = client.install_addon_with_checks(addons, addon);
	return_value = return_value && (res.outcome == addons_client::install_outcome::success);
}
```
which then wouldn't do an update for an outdated but existing add-on. Unless you mean getting rid of that if-check altogether?

Reply:
AFAICS, that if-check can be removed.

Reply:
Alright, updated.